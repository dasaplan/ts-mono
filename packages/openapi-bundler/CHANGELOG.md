# @dasaplan/openapi-bundler

## 0.0.7

### Patch Changes

- ef2415e: fixed toposort schemas when array items are allOf objects

## 0.0.6

### Patch Changes

- fix mergeAlOf order to follow default behaviour

## 0.0.5

### Patch Changes

- fix mergeAllOf order to follow default behaviour

## 0.0.3

### Patch Changes

- Update documentation

## 0.0.2

### Patch Changes

- handle dangling properties for schemas with allOf array.
