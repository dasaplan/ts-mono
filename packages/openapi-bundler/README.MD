# @dasaplan/openapi-bundler

- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
- âœ… Includes opinionated post-processing steps to transform the spec in a OpenApi 3.0.3 standard compliant way (see [spec-rules][spec-rules]).
- âœ… Is implemented against OpenApi 3.0.3.

## Getting started
### Prerequisite
- using npm
    ```shell
    # integrate in project
    npm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    npm i -g @dasaplan/openapi-bundler
    ```    
- using pnpm
    ```shell
    # integrate in project
    pnpm i --save-dev @dasaplan/openapi-bundler
    ```    
    ```shell
    # global installation
    pnpm i -g @dasaplan/openapi-bundler
    ```    

### usage
- Bundle an OpenapiSpec
  ```shell
  # assuming the root spec file is located at "$cwd/specs/generic/api.yml"
  # and assuming we want all generated files to find at "$cwd/out"
  openapi-bundler-cli bundle specs/generic/api.yml -o ./out/bundled-api.yml
  ```
- See help for more options
    ```shell
    openapi-bundler-cli bundle --help
    ```
  
## Features
- âœ… Configured wrapper of the standard [redocly](https://redocly.com/docs/cli/) bundler.
> note: Example specs are written in a compressed but valid way to reduce whitespace and facilitate an overview over the whole example.

### post processing
<details>
<summary>âœ… merge allOf subschemas</summary>

<details style="margin-left: 20px"><summary> motivation </summary>
Merging allOf arrays means we reduce complexity for generator and documentation tooling in order to facilitate consistency across tech stacks. 
<p>The easiest way would be to flat out every allOf so there resulting spec does not have the complexity at all. Yet, we need to find a middle ground because codegenerators may rely on identifiers like the discriminator to create object hierarchies.</p> 
We could just let every tool care for themselves how to handle their use-cases. However, from experience this will yield undefined behaviour for API consumer and lead to frustration because the default tooling does no yield results aligning with the specification.

```yaml
# schema declares $ref property but also other properties:
# most tools ignore everything besides $ref
# -> ðŸ”¥ unspecified behavior across tooling and use cases
RefSchema_with_dangling_properties:
  $ref: '#/A_Schema'
  description: "This is something I like to share!"
  title: RefSchema_with_dangling_properties

# schema declares allOf but declares further properties:
# most tool will merge siblings into array, but some may ignore them 
# -> ðŸ”¥ inconsistent behavior across tooling and use cases
AllOfSchema_with_dangling_properties:
  allOf: [ { $ref: '#/A_Schema'} ]
  properties: # sibling to allOf should be part of allOF
    name: { type: string }

# schema declares allOf with multiple elements:
# no problem for validators but code gen mostly try to merge them
# -> ðŸ”¥ merge behaviour may yield unexpected results
AllOfSchema_with_multi_elements:
  allOf:
    - $ref: '#/A_Schema',
    - $ref: '#/B_Schema',
    - properties: { foo: { type: string } }
```
</details>


<details style="margin-left: 20px">
<summary> src spec </summary>

````yaml
components:
  schemas:
    PetBase: 
      type: object,
      properties: { type: { type: string } }
    
    CatBase:  
      type: object,
      discriminator: { propertyName: catType },
      properties: 
        type: { type: string }, 
        catType: { type: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } 

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: SeamCat
    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - $ref: '#/components/CatBase'
        - title: ShortHairCat

    Dog: { allOf: [ { $ref: '#/components/PetBase' }, { title: Dog } ] }



````
</details>

<details style="margin-left: 20px">
<summary> resulting spec </summary>

````yaml
components:
  schemas:
    PetBase: { type: object, required: [type] discriminator: {propertyName: type},properties: { type: { type: string } } }

    SeamCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: SeamCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    ShortHairCat:
      allOf:
        - $ref: '#/components/PetBase'
        - { title: ShortHairCat, type: object, properties: { type: { catType: string, enum: [ 'SEAM', 'SHORT_HAIR' ] } } }

    Dog:
      allOf:
        - $ref: '#/components/PetBase'
        - title: Dog 
````
</details>
</details>



<details>
<summary>âœ… ensure discriminator values</summary>

- ensures that:
    - discriminator property type is of type string, and not enum.
        - most sophisticated tools can infer the values from the explicit or implicit discriminator mapping
    - polymorphism and inheritance can be explicitly inferred from the spec
        - every polymorph subschema defines the discriminator property and respective value
            - this is only necessary to extend the standard generator with templating
            - discriminator value is declared with <b>x-const</b> to avoid triggering compatibility layers for discriminator
        - any parent schema referenced from an allOf array does not define discriminator mapping
        - discriminator mapping only exists on schemas with a oneOf member

</details>
<details>
<summary>âœ… support multi schema usage in discriminator mapping (1:n discriminator value to mapping)</summary>

````yaml
Cat:
  type: object
  properties:
    type: 
      type: string
      x-const: [ 'SEAM', 'SHORT_HAIR' ]
    
MyOneOfSchema:
  oneOf: [{$ref: '#/Dog', {$ref: '#/Cat'}}]
  discriminator:
    propertyName: 'type'
    mapping:
      SEAM: "#/Cat"
      SHORT_HAIR: "#/Cat"
      Dog: "#/Dog"
````
</details>
<p></p>
<details>
<summary> <b style="color: #2c6fdf">ðŸ“œ complete example: openapi post-processing </b> </summary>

| examples | src                                 | bundled                                     | post-processed                                       |
|----------|-------------------------------------|---------------------------------------------|------------------------------------------------------|
| simple   | [link to src][simple-petstore-src]  | [link to bundled][simple-petstore-bundled]  | [link to post-processed][simple-petstore-processed]  |
| complex  | [link to src][complex-petstore-src] | [link to bundled][complex-petstore-bundled] | [link to post-processed][complex-petstore-processed] |

</details>

[complex-petstore-src]: ../openapi-specs/specs/pets-modular-complex/petstore-api.yml

[complex-petstore-bundled]: ./docs/examples/specs/complex-petstore-bundled.yml

[complex-petstore-processed]: ./docs/examples/specs/complex-petstore-post-processed.yml

[simple-petstore-src]: ../openapi-specs/specs/pets-simple/pets-api.yml

[simple-petstore-bundled]: ./docs/examples/specs/simple-petstore-bundled.yml

[simple-petstore-processed]: ./docs/examples/specs/simple-petstore-post-processed.yml

[spec-rules]: ./docs/spec-rules.md