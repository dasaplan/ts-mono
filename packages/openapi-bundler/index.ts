export {
  bundleOpenapi,
  createCommandBundle,
  createSpecProcessor,
  cleanObj,
  OpenApiBundled,
  parseOpenapi,
  Resolver,
  Transpiler,
  OaComponent,
  SchemaResolverContext,
  Schema,
} from "./src/index.js";
