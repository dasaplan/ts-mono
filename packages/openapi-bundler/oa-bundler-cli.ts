#!/usr/bin/env ts-node

import { Command } from "commander";
import * as process from "process";
import { createCommandBundle } from "./src/index.js";

const program = new Command();
createCommandBundle(program);

program.parse(process.argv);
