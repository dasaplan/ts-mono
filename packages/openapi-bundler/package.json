{
  "name": "@dasaplan/openapi-bundler",
  "description": "Opinionated bundler/merger for Openapi (Swagger)",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "version": "0.0.0",
  "keywords": [
    "openapi",
    "typescript",
    "ts",
    "zod",
    "codegen"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/dasaplan/ts-mono.git",
    "directory": "packages/openapi-bundler"
  },
  "scripts": {
    "prepack": "sort-package-json --quiet && clean-package",
    "postpack": "clean-package restore",
    "build": "pnpm rup",
    "rup": "tsup-node --dts",
    "lint": "eslint \"./src/*.ts\" index.ts oa-bundler-cli.ts",
    "clean": "rimraf dist",
    "test": "pnpm test:esm",
    "test:esm": "cross-env NODE_OPTIONS=--experimental-vm-modules jest",
    "test:esm:update": "cross-env NODE_OPTIONS=--experimental-vm-modules jest -u"
  },
  "bin": {
    "openapi-bundler-cli": "dist/oa-cli.js"
  },
  "files": ["dist"],
  "dependencies": {
    "commander": "^11.1.0",
    "json-schema-merge-allof": "^0.8.1",
    "jsonpointer": "^5.0.1",
    "openapi3-ts": "^4.3.3",
    "merge-json-schemas": "^1.0.0",
    "tslog": "^4.9.2",
    "lodash": "^4.17.21",
    "toposort": "^2.0.2"
  },
  "peerDependencies": {
    "@redocly/openapi-core": "^1.6.0"
  },
  "devDependencies": {
    "tsconfig": "workspace:*",
    "@types/json-schema-merge-allof": "^0.6.5",
    "@types/lodash":"^4.17.5",
    "@dasaplan/ts-sdk": "workspace:*"
  },
  "clean-package": {
    "remove": [
      "scripts",
      "devDependencies",
      "tsup",
      "clean-package",
      "dependencies.@dasaplan/ts-sdk"
    ]
  },
  "tsup": {
    "entry": [
      "index.ts",
      "oa-bundler-cli.ts"
    ],
    "noExternal": [
      "@dasaplan/ts-sdk",
      "tsconfig"
    ],
    "format": [
      "cjs",
      "esm"
    ],
    "tsconfig": "tsconfig.build.json",
    "clean": true,
    "sourcemap": true
  }
}
