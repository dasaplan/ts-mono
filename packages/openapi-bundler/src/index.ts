export { bundleOpenapi, OpenApiBundled, parseOpenapi } from "./bundle.js";
export { createCommandBundle } from "./commands.js";
export { createSpecProcessor } from "./post-process/index.js";
export { Transpiler, Schema } from "./transpiler/index.js";
export {
  Resolver,
  OaComponent,
  SchemaResolverContext,
  cleanObj,
} from "./resolver/index.js";
