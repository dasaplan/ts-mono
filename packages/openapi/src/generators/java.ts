import path from "path";
import process from "process";
import child_process from "node:child_process";
import { appLog } from "../logger.js";

export function generateJava(openapiSpec: string, out: string) {
  appLog.childLog(generateJava).info(`start generate:`, openapiSpec, out);
  const options = {
    enumUnknownDefaultCase: true,
  };
  const additionalProperties = Object.entries(options)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");
  const outDir = path.isAbsolute(out) ? out : path.resolve(process.cwd(), out, "java");
  child_process.execSync(
    `npx openapi-generator-cli generate -g spring --skip-validate-spec -i "${openapiSpec}" -o "${outDir}" --additional-properties=${additionalProperties}`
  );

  return outDir;
}
