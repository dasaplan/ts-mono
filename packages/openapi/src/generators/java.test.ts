import { Folder } from "@dasaplan/ts-sdk";
import { generateJava } from "./java.js";
import { bundleOpenapi, createSpecProcessor } from "@dasaplan/openapi-bundler";
import { resolveSpecPath } from "openapi-example-specs";
import path from "path";
import { describe, test, expect } from "vitest";

describe("Generator: java", () => {
  test("generate java", async () => {
    const spec = "pets-modular/pets-api.yml";
    const { outFile: bundled } = await bundleOpenapi(resolveSpecPath(spec), {
      postProcessor: createSpecProcessor({
        ensureDiscriminatorValues: true,
        mergeAllOf: true,
      }),
    });
    const outFolder = generateJava(bundled, "test/out/java");
    const files = Folder.of(outFolder).readAllFilesAsString();
    files.forEach((f) => expect(f.content).toMatchSnapshot(`java-${spec}-${path.basename(f.src)}`));
  });
});
