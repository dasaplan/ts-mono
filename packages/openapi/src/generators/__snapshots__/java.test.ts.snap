// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator: java generate java: java-pets-modular/pets-api.yml-.openapi-generator-ignore 1`] = `
"# OpenAPI Generator Ignore
# Generated by openapi-generator https://github.com/openapitools/openapi-generator

# Use this file to prevent files from being overwritten by the generator.
# The patterns follow closely to .gitignore or .dockerignore.

# As an example, the C# client generator defines ApiClient.cs.
# You can make changes and tell OpenAPI Generator to ignore just this file by uncommenting the following line:
#ApiClient.cs

# You can match any string of characters against a directory, file or extension with a single asterisk (*):
#foo/*/qux
# The above matches foo/bar/qux and foo/baz/qux, but not foo/bar/baz/qux

# You can recursively match patterns against a directory, file or extension with a double asterisk (**):
#foo/**/qux
# This matches foo/bar/qux, foo/baz/qux, and foo/bar/baz/qux

# You can also negate patterns with an exclamation (!).
# For example, you can ignore all files in a docs folder with the file extension .md:
#docs/*.md
# Then explicitly reverse the ignore rule for a single file:
#!docs/README.md
"
`;

exports[`Generator: java generate java: java-pets-modular/pets-api.yml-README.md 1`] = `
"# OpenAPI generated server

Spring Boot Server

## Overview
This server was generated by the [OpenAPI Generator](https://openapi-generator.tech) project.
By using the [OpenAPI-Spec](https://openapis.org), you can easily generate a server stub.
This is an example of building a OpenAPI-enabled server in Java using the SpringBoot framework.


The underlying library integrating OpenAPI to Spring Boot is [springdoc](https://springdoc.org).
Springdoc will generate an OpenAPI v3 specification based on the generated Controller and Model classes.
The specification is available to download using the following url:
http://localhost:8080/v3/api-docs/

Start your server as a simple java application

You can view the api documentation in swagger-ui by pointing to
http://localhost:8080/swagger-ui.html

Change default port value in application.properties"
`;

exports[`Generator: java generate java: java-pets-modular/pets-api.yml-pom.xml 1`] = `
"<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.openapitools</groupId>
    <artifactId>openapi-spring</artifactId>
    <packaging>jar</packaging>
    <name>openapi-spring</name>
    <version>1.0.0</version>
    <properties>
        <java.version>1.8</java.version>
        <maven.compiler.source>\${java.version}</maven.compiler.source>
        <maven.compiler.target>\${java.version}</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <springdoc.version>1.6.14</springdoc.version>
        <swagger-ui.version>5.3.1</swagger-ui.version>
    </properties>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.7.15</version>
        <relativePath/> <!-- lookup parent from repository -->
    </parent>
    <build>
        <sourceDirectory>src/main/java</sourceDirectory>
        <plugins>
            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
            </plugin>
        </plugins>
    </build>
    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-commons</artifactId>
        </dependency>
          <!--SpringDoc dependencies -->
        <dependency>
            <groupId>org.springdoc</groupId>
            <artifactId>springdoc-openapi-ui</artifactId>
            <version>\${springdoc.version}</version>
        </dependency>
        <!-- @Nullable annotation -->
        <dependency>
            <groupId>com.google.code.findbugs</groupId>
            <artifactId>jsr305</artifactId>
            <version>3.0.2</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.dataformat</groupId>
            <artifactId>jackson-dataformat-yaml</artifactId>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jsr310</artifactId>
        </dependency>
        <dependency>
            <groupId>org.openapitools</groupId>
            <artifactId>jackson-databind-nullable</artifactId>
            <version>0.2.6</version>
        </dependency>
        <!-- Bean Validation API support -->
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-validation</artifactId>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>
</project>
"
`;
